test/variational/test_qgt.py

```diff
@@ -0,0 +1,43 @@
+import netket as nk
+import jax
+from jax.flatten_util import ravel_pytree
+import numpy as np
+import netket.nn as nknn
+
+hi =nk.hilbert.Spin(0.5, 4)
+ma = nk.models.RBM(alpha=1, dtype=float,kernel_init=nknn.initializers.normal(0.5))
+sa = nk.sampler.ExactSampler(hi)
+
+n_samples=1600000
+vs = nk.vqs.MCState(sa, ma,n_samples=n_samples)
+
+qgt = vs.quantum_geometric_tensor()
+
+
+all_states=jax.numpy.asarray(hi.all_states())
+
+def NormPsi(pars):
+    return jax.numpy.sqrt(jax.numpy.sum(jax.numpy.abs(jax.numpy.exp(ma.apply(pars,all_states)))**2))
+
+
+def Psi(pars,x):
+    return jax.numpy.exp(ma.apply(pars,x))/NormPsi(pars)
+
+def GradPsiSingle(pars,x):
+    return ravel_pytree(jax.grad(Psi,argnums=0)(pars,x))[0]
+
+
+def GradPsi(pars,x):
+    return jax.vmap(GradPsiSingle,(None,0))(pars,x)
+
+
+
+PsiVec=Psi(vs.variables,all_states)
+GVec=GradPsi(vs.variables,all_states)
+
+#The exact qgt is <GradPsi|GradPsi> -<GradPsi|Psi><Psi|GradPsi>
+qgt_ex=(GVec.conjugate().transpose()@GVec) -(GVec.conjugate().transpose()@PsiVec)*(PsiVec.conjugate().transpose()@GVec)
```

Comment:
I don't think that matters, the centring term should vanish if the wave function is normalised. I also think it should for a holomorphic C→C function too (e.g. 0 = d/dz* <Ψ|Ψ> = <dΨ/dz*|Ψ> because the second term has no z* derivative), so it would be good to check whether it actually does. 

Also, both `(GVec.conjugate().transpose()@PsiVec)` and `(PsiVec.conjugate().transpose()@GVec)` are just 1D vectors, so multiplying them like this gives a vector that broadcasts on the matrix, rather than the outer product matrix. However, this shouldn't matter as they're supposed to vanish, but there's definitely something fishy going on with this line

Reply:
erm yes, you are right, there should be an outer product....

Reply:
I just checked and `(GVec.conjugate().transpose()@PsiVec)` is numerically zero only for `dtype=float`.
It is not for `dtype=complex`.

Reply:
I think that also for R->C it is non zero, now I fixed the outer product and R->R and R->C both seem to work correctly, I still have issues with C->C

Reply:
yeah i can reproduce...

this is very curious.

Reply:
`Psi(pars,x)` is **not** a holomorphic function of `pars` because we divide through with `NormPsi` which is clearly not holomorphic, so `grad` will produce garbage. Taking the gradients of the unnormalised wave function and normalising afterwards fixes the error: https://gist.github.com/attila-i-szabo/fe35c1e874d9b6902d60812a6cea9db5

Reply:
I love you

Reply:
Great solve!